src/internet_identity/src/storage/anchor.rs

```diff
@@ -0,0 +1,391 @@
+use candid::{CandidType, Deserialize, Principal};
+use internet_identity_interface::archive::DeviceDataWithoutAlias;
+use internet_identity_interface::*;
+use std::{fmt, iter};
+
+#[cfg(test)]
+mod tests;
+
+/// Internal representation of the anchor.
+/// The anchor has limited visibility for the constructor to make sure it is loaded from storage.
+/// The devices can only be modified by the exposed functions which keeps invariant checking local
+/// to this module.
+///
+/// After the upcoming stable memory migration, new fields can be added to this struct.
+#[derive(Clone, Debug, Default, CandidType, Deserialize, Eq, PartialEq)]
+pub struct Anchor {
+    devices: Vec<Device>,
+}
+
+impl From<DeviceData> for Device {
+    fn from(device_data: DeviceData) -> Self {
+        Self {
+            pubkey: device_data.pubkey,
+            alias: device_data.alias,
+            credential_id: device_data.credential_id,
+            purpose: device_data.purpose,
+            key_type: device_data.key_type,
+            protection: device_data.protection,
+        }
+    }
+}
+
+impl From<Device> for DeviceData {
+    fn from(device: Device) -> Self {
+        Self {
+            pubkey: device.pubkey,
+            alias: device.alias,
+            credential_id: device.credential_id,
+            purpose: device.purpose,
+            key_type: device.key_type,
+            protection: device.protection,
+        }
+    }
+}
+
+impl From<Device> for DeviceDataWithoutAlias {
+    fn from(device_data: Device) -> Self {
+        Self {
+            pubkey: device_data.pubkey,
+            credential_id: device_data.credential_id,
+            purpose: device_data.purpose,
+            key_type: device_data.key_type,
+            protection: device_data.protection,
+        }
+    }
+}
+
+impl Anchor {
+    /// Creation of new anchors is restricted in order to make sure that the device checks are
+    /// not accidentally bypassed.
+    pub(super) fn new() -> Anchor {
+        Self { devices: vec![] }
+    }
+
+    /// Creation of new anchors is restricted in order to make sure that the device checks are
+    /// not accidentally bypassed.
+    pub(super) fn from_devices(devices: Vec<Device>) -> Anchor {
+        // We do _not_ check invariants here, because there might be anchors that do not fulfill
+        // the invariants still stored in stable memory (e.g. anchors with multiple recovery phrases).
+        Self { devices }
+    }
+
+    pub fn add_device(&mut self, device: Device) -> Result<(), AnchorError> {
+        if self
+            .devices
+            .iter()
+            .find(|e| e.pubkey == device.pubkey)
+            .is_some()
+        {
+            return Err(AnchorError::DuplicateDevice {
+                device_key: device.pubkey,
+            });
+        }
+        check_device(&device)?;
+        check_invariants(&self.devices.iter().chain(iter::once(&device)).collect())?;
```

Comment:
I think I would have understood the intention behind these functions a little easier if they were named `check_device_invariants` and `check_anchor_invariants`.

Reply:
good suggestion 👍 --> renamed